File: src/Polls/LJPollTextEntryQuestion.m
//
//  LJPollTextEntryQuestion.m
//  PollTest
//
//  Created by Fraser Speirs on Fri Mar 14 2003.
//  Copyright (c) 2003 Connected Flow. All rights reserved.
//

#import "LJPollTextEntryQuestion.h"

#define kPollTextSize @"LJPollTextSize"
#define kPollTextLength @"LJPollTextLength"

@implementation LJPollTextEntryQuestion
@synthesize size;
@synthesize maxLength;

- (instancetype)init
{
    return [self initWithSize:10 maxLength:10];
}

- (instancetype)initWithSize:(NSInteger)theSize maxLength:(NSInteger)theLength
{
    if (self = [super init]) {
        self.question = @"New Text Question";
        size = theSize;
        maxLength = theLength;
    }
    return self;
}

+ (LJPollTextEntryQuestion *)textEntryQuestionWithSize: (NSInteger)theSize maxLength: (NSInteger)theLength
{
    LJPollTextEntryQuestion *teQ = [[LJPollTextEntryQuestion alloc] initWithSize:theSize maxLength:theLength];

    return teQ;
}

- (NSString *)htmlRepresentation
{
    return [NSString stringWithFormat: @"<lj-pq type=\"text\" size=\"%ld\" maxlength=\"%ld\">%@</lj-pq>", (long)size, (long)maxLength, self.question];
}

#pragma mark Memento Pattern
- (NSDictionary *) memento
{
    NSDictionary *dict = @{kLJPollQuestionKey   : self.question,
                           kPollTextSize        : @(size),
                           kPollTextLength      : @(maxLength)};

    return dict;
}

- (void) restoreFromMemento: (NSDictionary *)memento
{
    self.question = memento[kLJPollQuestionKey];
    self.size = [memento[kPollTextSize] integerValue];
    self.maxLength = [memento[kPollTextLength] integerValue];
}

#pragma mark NSCoding
- (void)encodeWithCoder:(NSCoder *)encoder
{
    [super encodeWithCoder: encoder];
    [encoder encodeInteger: maxLength forKey: kPollTextLength];
    [encoder encodeInteger: size forKey: kPollTextSize];
}

- (instancetype)initWithCoder:(NSCoder *)decoder
{
    if (self = [super initWithCoder: decoder]) {
        maxLength = [decoder decodeIntegerForKey: kPollTextLength];
        size = [decoder decodeIntegerForKey: kPollTextSize];
    }
    return self;
}
@end

File: src/Polls/LJPollTextEntryQuestion.swift
//
//  LJPollTextEntryQuestion.swift
//  Xjournal
//
//  Created by C.W. Betts on 1/24/15.
//
//

import Foundation

private let kPollTextSize = "LJPollTextSize"
private let kPollTextLength = "LJPollTextLength"

///This is a subclass representing a text box question
@objc(LJPollTextEntryQuestion) final class PollTextEntryQuestion: PollQuestion {
	/// Get and set the size property
	var size: Int
	/// Get and set the length
	var maxLength: Int
	
	/// Returns a text question with the given size and length
	/// and a default question
	init(size theSize: Int, maxLength theLength: Int) {
		size = theSize
		maxLength = theLength
		
		super.init()
		question = "New Text Question"
	}
	
	/// Returns a text question a size of 10, a length of 10,
	/// and a default question
	override convenience init() {
		self.init(size: 10, maxLength: 10)
	}
	
	/// Get the HTML representation
	override var htmlRepresentation: String {
		return "<lj-pq type=\"text\" size=\"\(size)\" maxlength=\"\(maxLength)\">\(question)</lj-pq>"
	}
	
	// MARK: - memento
	override var memento: [String: Any] {
		return [kLJPollQuestionKey: question,
			kPollTextSize: size,
			kPollTextLength: maxLength]
	}
	
	override func restore(fromMemento amemento: [String: Any]) {
		question = amemento[question] as! String
		size = amemento[kPollTextSize] as! Int
		maxLength = amemento[kPollTextLength] as! Int
	}

	// MARK: - NSCoding
	override func encode(with aCoder: NSCoder) {
		super.encode(with: aCoder)
		aCoder.encode(size, forKey: kPollTextSize)
		aCoder.encode(maxLength, forKey: kPollTextLength)
	}
	
	required init?(coder aDecoder: NSCoder) {
		size = aDecoder.decodeInteger(forKey: kPollTextSize)
		maxLength = aDecoder.decodeInteger(forKey: kPollTextLength)
		
		super.init(coder: aDecoder)
	}
	
	// MARK: -
	class var keyPathsForValuesAffectingHtmlRepresentation: Set<String> {
		return Set(["size", "maxLength", "question"])
	}
}
